components/patient-data/impl/src/main/java/org/phenotips/data/internal/EmptyGenotypeObjectsRemover.java

```diff
@@ -74,54 +74,29 @@ public void onEvent(Event event, Object source, Object data)
         List<BaseObject> geneXWikiObjects = doc.getXObjects(Gene.GENE_CLASS);
         List<BaseObject> variantXWikiObjects = doc.getXObjects(VARIANT_CLASS_REFERENCE);
 
-        if ((geneXWikiObjects == null || geneXWikiObjects.isEmpty())
-            && (variantXWikiObjects != null && !variantXWikiObjects.isEmpty())) {
-            // delete all variants
+        if (geneXWikiObjects == null || geneXWikiObjects.isEmpty()) {
+            // Delete all variants
             doc.removeXObjects(VARIANT_CLASS_REFERENCE);
             return;
         }
 
-        // get list of patient genes
-        Set<String> genes = new HashSet<>();
-        for (BaseObject geneObject : geneXWikiObjects) {
-            if (geneObject == null) {
-                continue;
-            }
-
-            BaseStringProperty field = (BaseStringProperty) geneObject.getField(GENE_KEY);
-            // remove gene object from the document if key "gene" field is empty
-            if (field == null || StringUtils.isEmpty(field.getValue())) {
-                doc.removeXObject(geneObject);
-                continue;
-            }
-
-            genes.add(field.getValue());
-        }
+        // Remove gene object from the document if key "gene" field is empty
+        geneXWikiObjects.stream().filter(Objects::nonNull)
+            .filter(gene -> StringUtils.isBlank(gene.getStringValue(GENE_KEY)))
+            .forEach(gene -> doc.removeXObject(gene));
 
         if (variantXWikiObjects == null || variantXWikiObjects.isEmpty()) {
-            // nothing to remove
+            // Nothing else to remove
```

Comment:
we can move line 75 to line 87 since we don't need variants before that point but can exit function.